Implement index view

kjcioffi opened the pull request:
- Refactored requirements into a folder for improved dependency management.
- Implemented business logic to route movies playing in theaters to a view.
- Implemented a view that populates movies playing in theaters using previously mentioned business logic written in movie_api_utils.py

kjcioffi committed:
Initial commit of TestSuite for Index view.
   - Added requests to project requirements.
   - Re-structured test file layout.

kjcioffi committed:
Wired index view to app and project level urls.py

kjcioffi committed:
First test in Index view test suite.
   - Refactored API info out of the test suite into a separate module allowing for re-usability across tests and standard development modules.
   - Initial test providing coverage for context item check.

kjcioffi committed:
Development completed for initial test.

kjcioffi committed:
Improved test by checking for Response instance.

kjcioffi committed:
Enabled django-debug-toolbar.
   - Run default migrations and configured the respective files to ensure the toolbar functions.
   - Refactored project and dev requirements into a directory.

kjcioffi committed:
Formatted core/settings.py for testing.
   - Django toolbar was causing issues with executing tests.
   - Implemented logic that prevents the toolbar from being installed if tests are being executed.

kjcioffi committed:
Index now returns a dictionary object rather than a Response from requests.

kjcioffi committed:
Removed debug from MovieApiUtils object, the argument is not require.

kjcioffi committed:
API_KEY default set for MovieApiUtils.

kjcioffi committed:
Re-factored function based index view into a class based view named MovieListView.
   - Template written for the view to populate data.

kjcioffi committed:
Refactored MovieApiUtils to filter movies by a beginning date.

kjcioffi committed:
Removing index.html due to rename of the view.

kjcioffi committed:
Adding base.html parent template.

kjcioffi committed:
Refactored test suite to include rename changes. Also did some dependency and white space cleanup.

kjcioffi committed:
Remove duplicate debug middleware from MIDDLEWARE list in settings.py.

kjcioffi committed:
Docstring added for MovieApiUtils class (movie_api_utils module).

ryaustin commented on movies/tests/views/test_movie_list_view.py:
```diff
@@ -0,0 +1,13 @@
+        self.movies_in_context = self.response.context["movies"]
+
+    def test_movies_in_context(self):
+        self.assertTrue(self.movies_in_context)
```
To make this test more robust, it might be worthwhile to ensure there are no errors. 
When I ran this test it passed successfully - which I expected to fail since I don't have the API key set yet.
Here is the resolve of self.movies_in_context: `{'status_code': 7, 'status_message': 'Invalid API key: You must be granted a valid key.', 'success': False}`

So how can you 1: make the test more robust? 
Does it make sense to test a scenario where there is no API key (success: False)?

ryaustin commented on core/settings.py:
```diff
@@ -62,6 +72,18 @@
     "django.middleware.clickjacking.XFrameOptionsMiddleware",
 ]
 
+if DEBUG_TOOLBAR_ENABLED:
```
This is well implemented. 
To make it even more robust I suggest placing "debug_toolbar" and the debug_toolbar middleware string in a constant and using it to increase readability & maintainability `DEBUG_TOOLBAR = "debug_toolbar"`

ryaustin commented on core/settings.py:
```diff
@@ -62,6 +72,18 @@
+            mw
+            for mw in MIDDLEWARE
+            if "debug_toolbar.middleware.DebugToolbarMiddleware" not in mw
+        ]
```
here we aren't checking if the debug_toolbar.middleware is in the middleware before removing it.
I guess it can be pretty safely assumed that it is.

ryaustin commented on core/settings.py:
```diff
@@ -36,7 +37,14 @@
-ALLOWED_HOSTS = []
+DEBUG_TOOLBAR_ENABLED = DEBUG and "test" not in sys.argv
+
+ALLOWED_HOSTS = ["*"]
```
I recommend managing allowed_hosts explicitly from day 1. This reduces headache and troubleshooting when it's time to lockdown and deploy. Right now it's localhost and 127.x.x. later you might use an environment variable to set ALLOWED_HOSTS appropriately.
 `ALLOWED_HOSTS = env.list("ALLOWED_HOSTS", default=["localhost"])`

ryaustin commented on core/urls.py:
```diff
@@ -16,8 +16,10 @@
 
 urlpatterns = [
+    path("__debug__/", include("debug_toolbar.urls")),
+    path("", include("movies.urls", namespace="movies")),
```
`namespaces are a honking great idea`

ryaustin commented on movies/movie_api_utils.py:
Looking great here. Consider exception handling when using requests to manage potential issues like network errors or invalid responses. Also, exception handling is considered foundational knowledge for the pcpd-p cert.

ryaustin commented on movies/views.py:
here we can make several small improvements for robustness, readability and maintainability of your code:
- error handling when calling requests - you might consider making this call to requests a part of your MovieApiUtils class so that you are not constantly dealing with exception handling in your views.
- Type hints, It’s generally recommended to use Dict from typing for clarity and consistency, especially when sharing code with others. small 'd' still works but it's one of those things that might cause a head-scratcher if someone else with a older version of python runs your code.
- You want to check if your response is valid before requesting JSON. here you can use response.raise_for_status() because It helps ensure that your code gracefully handles situations where an HTTP request fails.

ryaustin reviewed (changes requested):
Great submission here Kevin. Your imports are clean, dependencies are expertly managed and the code just works.
Most of my commentary is around the unhappy times when things that you expect to work don't (such as the API being down).

I recommend making some specific changes that I will breakout as issues.

kjcioffi committed:
Implemented authentication method in MovieApiUtils with a newly associated test suite.

kjcioffi committed:
Implemented exception handling into MovieApiUtils class w/ associated test coverage.
   - Incorporated authentication checks into get_movies_now_playing() method.
   - Test coverage added to handle issues with bad or dsyfunctional API keys.

ryaustin commented:
Hi Kevin, please push the latest changes made on 12-Jun-24 along with the working tests so that we can merge this PR.

kjcioffi committed:
Refactored MovieApiUtils test suite for improved test coverage.

kjcioffi committed:
Implemented a wrapper for current and all future methods in the MovieApiUtils module.

kjcioffi committed:
Refactored get_movies_now_playing to call the wrapper with exception handling.

kjcioffi committed:
Refactored authenticate method to call the wrapper with exception handling.

kjcioffi committed:
Changed type hint return value of _get() to better repesent what is being returned.

kjcioffi committed:
Added type hints and a method docstring for get_movies_now_playing().

kjcioffi committed:
Added type hints and a method docstring for authenticate() in MovieApiUtil module.

kjcioffi committed:
Import module for type hints.

kjcioffi committed:
Implemented error handling in MovieListView(index).

kjcioffi committed:
Refactored ALLOWED_HOSTS in settings.py to retrieve values from .env file.

kjcioffi committed:
Created constants in core/settings.py for the debug toolbar.

kjcioffi committed:
No need for sqlite.

kjcioffi commented:
Ready for a final (hopefully) review

cc @ryaustin

ryaustin reviewed (approved):
HI Kevin, great job with the changes.
My tests failed but I can tell that it has to do with the API key not being authenticated. This is something we can look at together but does not prevent this code from being merged in as that can be reviewed as a separate issue. Great job.

ryaustin commented:
Go ahead and merge whenever you're ready.
